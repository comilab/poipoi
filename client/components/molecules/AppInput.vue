<template lang="pug">
ValidationProvider(
  :vid="vid",
  :name="name || label",
  :rules="rules",
  v-slot="{ classes, errors }",
  :immediate="immediate",
  slim
)
  .flex.flex-col(
    :class="classes",
    :data-testid="`input-${name || label || vid}`"
  )
    AppLabel(
      v-if="hasLabel",
      :for="id",
      data-testid="label"
    )
      slot(name="label") {{ label }}
    template(v-if="type === 'checkbox'")
      label.cursor-pointer(data-testid="input")
        input.mr-2(
          type="checkbox",
          v-model="innerValue",
          :value="value",
          :name="name || label || vid || checkboxLabel",
          v-bind="$attrs"
        )
        span {{ checkboxLabel }}
    template(v-else-if="type === 'checkboxgroup'")
      .flex.flex-wrap.-mx-4.-my-2(data-testid="input")
        label.cursor-pointer.mx-4.my-2(v-for="(option, i) in options", :key="i")
          input.mr-2(
            type="checkbox",
            v-model="innerValue",
            :value="option.value",
            :disabled="option.disabled",
            :name="name || label || vid",
            v-bind="$attrs"
          )
          span.whitespace-nowrap {{ option.label }}
    template(v-else-if="type === 'radio'")
      .flex.flex-wrap.-mx-4.-my-2(data-testid="input")
        label.cursor-pointer.mx-4.my-2(v-for="(option, i) in options", :key="i")
          input.mr-2(
            type="radio",
            v-model="innerValue",
            :value="option.value",
            :disabled="option.disabled",
            :name="name || label || vid",
            v-bind="$attrs"
          )
          span.whitespace-nowrap {{ option.label }}
    template(v-else-if="type === 'textarea'")
      textarea.border-b.focus_border-blue-400.focus_outline-none.transition-colors.duration-300.p-2(
        :id="id",
        ref="textarea",
        v-model="innerValue",
        :name="name || label || vid",
        v-bind="$attrs",
        @keydown="adjustHeight",
        data-testid="input"
      )
    template(v-else)
      input.border-b.focus_border-blue-400.focus_outline-none.transition-colors.duration-300.bg-white.p-2(
        :id="id",
        :type="type",
        v-model="innerValue",
        :name="name || label || vid",
        v-bind="$attrs",
        data-testid="input"
      )
    .flex.flex-col.md_flex-row.flex-col-reverse.justify-between.text-sm
      div(
        v-if="!hideErrors",
        data-testid="errors"
      )
        .mt-1(
          v-for="(error, i) in errors"
        ) {{ error }}
      .text-gray-500.text-right.mt-1(
        v-if="rules.max",
        data-testid="counter"
      ) {{ innerValue.length }} / {{ rules.max }}
</template>

<script lang="ts">
import { defineComponent, ref, computed, onMounted } from 'nuxt-composition-api'
import { PropType } from '@vue/composition-api'
import dayjs from 'dayjs'

import useUniqId from '~/composables/use-uniq-id'
import AppLabel from '~/components/atoms/AppLabel.vue'

type Option = {
  label: string
  value: any
  disabled: boolean
}

export default defineComponent({
  components: {
    AppLabel
  },
  props: {
    value: {
      required: false,
      default: null
    },
    type: {
      type: String,
      required: true
    },
    label: {
      type: String,
      required: false,
      default: null
    },
    vid: {
      type: String,
      required: false,
      default: null
    },
    name: {
      type: String,
      required: false,
      default: null
    },
    options: {
      type: Array as PropType<Option[]>,
      required: false,
      default: () => []
    },
    checkboxLabel: {
      type: String,
      required: false,
      default: null
    },
    rules: {
      type: [String, Object] as PropType<any>,
      required: false,
      default: () => { return {} }
    },
    immediate: {
      type: Boolean,
      required: false,
      default: false
    },
    hideErrors: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  setup (props, context) {
    const { id } = useUniqId('input')

    const textarea = ref<HTMLTextAreaElement>()

    const innerValue = computed({
      get: () => {
        if (props.type === 'datetime-local' && props.value) {
          return dayjs(props.value as any).format('YYYY-MM-DDTHH:mm')
        }
        return props.value
      },
      set: (value) => {
        if (props.type === 'datetime-local' && value) {
          context.emit('input', dayjs(value).utc().toDate())
        } else {
          context.emit('input', value)
        }
        adjustHeight()
      }
    })

    const hasLabel = computed(() => {
      return props.label || context.slots.label
    })

    const adjustHeight = async () => {
      if (textarea.value) {
        textarea.value.style.height = 'auto'
        await context.root.$nextTick()
        const height = Math.min(Math.max(75, textarea.value.scrollHeight), 400)
        textarea.value.style.height = `${height}px`
      }
    }

    onMounted(() => {
      if (textarea.value) {
        adjustHeight()
      }
    })

    return {
      id,
      textarea,
      innerValue,
      hasLabel,
      adjustHeight
    }
  }
})
</script>

<style lang="postcss" scoped>
.invalid {
  @apply text-red-700;

  & input,
  & textarea {
    @apply text-black border-red-700;
  }

  & div {
    @apply text-red-700;
  }
}
</style>
