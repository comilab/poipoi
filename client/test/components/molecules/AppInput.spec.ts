import { render, RenderOptions, fireEvent, within, waitFor } from '@testing-library/vue'
import { defineComponent } from 'nuxt-composition-api'
import dayjs from 'dayjs'

import AppInput from '~/components/molecules/AppInput.vue'

describe('components/molecules/AppInput', () => {
  let options: RenderOptions<AppInput>
  let props: any

  beforeEach(() => {
    props = {
      type: 'textd'
    }

    options = {
      props
    }
  })

  test('renders correctly', () => {
    const { container } = render(AppInput, options)

    expect(container).toMatchSnapshot()
  })

  describe('props', () => {
    describe('value', () => {
      describe('get', () => {
        test('v-modelとしてセット', () => {
          props.value = 'foo'
          const { container, getByDisplayValue } = render(AppInput, options)

          expect(container).toMatchSnapshot()
          getByDisplayValue('foo')
        })

        test('typeがdatetime-localだった場合は日付文字列にフォーマット', () => {
          props.type = 'datetime-local'
          props.value = dayjs('2020-01-01 00:00').toDate()
          const { container, getByDisplayValue } = render(AppInput, options)

          expect(container).toMatchSnapshot()
          getByDisplayValue('2020-01-01T00:00')
        })
      })

      describe('set', () => {
        test('inputイベントを発火', async () => {
          const { getByTestId, emitted } = render(AppInput, options)
          await fireEvent.update(getByTestId('input'), 'foo')

          expect(emitted().input).toHaveLength(1)
          expect(emitted().input[0]).toStrictEqual(['foo'])
        })

        test('typeがdatetime-localだった場合はDate型に変換', async () => {
          props.type = 'datetime-local'
          props.value = dayjs('2020-01-01 00:00').toDate()
          const { getByTestId, emitted } = render(AppInput, options)
          await fireEvent.update(getByTestId('input'), '2020-01-02 00:00')

          expect(emitted().input[0][0]).toBeInstanceOf(Date)
          expect(dayjs(emitted().input[0][0]).isSame('2020-01-02 00:00'))
        })

        test('typeがdatetime-localかつ値が空だった場合はそのまま渡す', async () => {
          props.type = 'datetime-local'
          props.value = dayjs('2020-01-01 00:00').toDate()
          const { getByTestId, emitted } = render(AppInput, options)
          await fireEvent.update(getByTestId('input'), '')

          expect(emitted().input[0][0]).toBe('')
        })

        test('typeがtextareaだった場合は高さを調節', async () => {
          props.type = 'textarea'
          const { getByTestId, queryByTestId } = render(AppInput, options)
          const height = queryByTestId('input')!.style.height

          await fireEvent.update(getByTestId('input'), 'foo')

          expect(queryByTestId('input')!.style.height).not.toBe(height)
        })
      })
    })

    describe('type', () => {
      test('checkbox', () => {
        props.type = 'checkbox'
        const { container } = render(AppInput, options)

        expect(container).toMatchSnapshot()
      })

      test('checkboxgroup', () => {
        props.type = 'checkboxgroup'
        const { container } = render(AppInput, options)

        expect(container).toMatchSnapshot()
      })

      test('radio', () => {
        props.type = 'radio'
        const { container } = render(AppInput, options)

        expect(container).toMatchSnapshot()
      })

      test('textarea', () => {
        props.type = 'textarea'
        const { container } = render(AppInput, options)

        expect(container).toMatchSnapshot()
      })
    })

    describe('label', () => {
      test('ラベルとして表示', () => {
        props.label = 'label'
        const { container, getByLabelText } = render(AppInput, options)

        expect(container).toMatchSnapshot()
        getByLabelText('label')
      })

      test('空の場合はlabel要素自体を描画しない', () => {
        const { container, queryByTestId } = render(AppInput, options)

        expect(container).toMatchSnapshot()
        expect(queryByTestId('label')).not.toBeInTheDocument()
      })

      test('nameがない場合はlabelをValidationProviderのnameに渡す', () => {
        let validationProviderProps: any
        options.stubs = {
          ValidationProvider: defineComponent({
            props: ['name'],
            setup (props) {
              validationProviderProps = props
            },
            template: '<div />'
          })
        }
        props.label = 'foo'
        render(AppInput, options)

        expect(validationProviderProps.name).toBe('foo')
      })
    })

    describe('vid', () => {
      test('ValidationProviderに渡す', () => {
        let validationProviderProps: any
        options.stubs = {
          ValidationProvider: defineComponent({
            props: ['vid'],
            setup (props) {
              validationProviderProps = props
            },
            template: '<div />'
          })
        }
        props.vid = 'foo'
        render(AppInput, options)

        expect(validationProviderProps.vid).toBe('foo')
      })
    })

    describe('name', () => {
      test('ValidationProviderのnameに渡す', () => {
        let validationProviderProps: any
        options.stubs = {
          ValidationProvider: defineComponent({
            props: ['name'],
            setup (props) {
              validationProviderProps = props
            },
            template: '<div />'
          })
        }
        props.name = 'foo'
        render(AppInput, options)

        expect(validationProviderProps.name).toBe('foo')
      })
    })

    describe('options', () => {
      test('checkboxgroupで正常に描画される', () => {
        props.options = [
          { label: 'label', value: 'value' }
        ]
        props.type = 'checkboxgroup'
        const { getByTestId } = render(AppInput, options)

        within(getByTestId('input')).getByLabelText('label')
        within(getByTestId('input')).getByDisplayValue('value')
      })

      test('radioで正常に描画される', () => {
        props.options = [
          { label: 'label', value: 'value' }
        ]
        props.type = 'radio'
        const { getByTestId } = render(AppInput, options)

        within(getByTestId('input')).getByLabelText('label')
        within(getByTestId('input')).getByDisplayValue('value')
      })
    })

    describe('checkboxLabel', () => {
      test('checkboxのラベルとして描画', () => {
        props.checkboxLabel = 'checkbox-label'
        props.type = 'checkbox'
        const { getByTestId } = render(AppInput, options)

        within(getByTestId('input')).getByLabelText('checkbox-label')
      })
    })

    describe('rules', () => {
      test('ValidationProviderのrulesに渡す', () => {
        let validationProviderProps: any
        options.stubs = {
          ValidationProvider: defineComponent({
            props: ['rules'],
            setup (props) {
              validationProviderProps = props
            },
            template: '<div />'
          })
        }
        props.rules = 'required'
        render(AppInput, options)

        expect(validationProviderProps.rules).toBe('required')
      })

      test('maxがあれば文字数カウンターを表示', () => {
        props.value = ''
        props.rules = { max: 300 }
        const { container, getByTestId } = render(AppInput, options)

        expect(container).toMatchSnapshot()
        getByTestId('counter')
      })
    })

    describe('immediate', () => {
      test('ValidationProviderのrulesに渡す', () => {
        let validationProviderProps: any
        options.stubs = {
          ValidationProvider: defineComponent({
            props: ['immediate'],
            setup (props) {
              validationProviderProps = props
            },
            template: '<div />'
          })
        }
        props.immediate = true
        render(AppInput, options)

        expect(validationProviderProps.immediate).toBeTruthy()
      })
    })

    describe('hideErrors', () => {
      test('trueならエラーがあっても表示しない', async () => {
        props.rules = 'required'
        const { getByTestId, updateProps, container, queryByTestId } = render(AppInput, options)
        getByTestId('errors')

        await updateProps({ hideErrors: true })

        expect(container).toMatchSnapshot()
        expect(queryByTestId('errors')).not.toBeInTheDocument()
      })
    })
  })

  describe('slots', () => {
    describe('label', () => {
      test('ラベルとして表示', () => {
        options.slots = { label: 'label' }
        const { container, getByLabelText } = render(AppInput, options)

        expect(container).toMatchSnapshot()
        getByLabelText('label')
      })
    })
  })

  describe('adjustHeight', () => {
    let scrollHeight: number

    beforeEach(() => {
      scrollHeight = 100
      jest.spyOn(HTMLTextAreaElement.prototype, 'scrollHeight', 'get')
        .mockImplementation(() => scrollHeight)
      props.type = 'textarea'
    })

    test('マウント時に実行', async () => {
      const { queryByTestId } = render(AppInput, options)

      await waitFor(() => {
        expect(queryByTestId('input')!.style.height).toBe('100px')
      })
    })

    test('高さの最小値は75px', async () => {
      scrollHeight = 74
      const { queryByTestId } = render(AppInput, options)

      await waitFor(() => {
        expect(queryByTestId('input')!.style.height).toBe('75px')
      })
    })

    test('高さの最大値は400px', async () => {
      scrollHeight = 401
      const { queryByTestId } = render(AppInput, options)

      await waitFor(() => {
        expect(queryByTestId('input')!.style.height).toBe('400px')
      })
    })
  })
})
