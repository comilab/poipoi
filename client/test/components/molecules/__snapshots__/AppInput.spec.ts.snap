// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/AppInput props hideErrors trueならエラーがあっても表示しない 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine pending required"
    data-testid="input-null"
  >
    <!---->
    <input
      class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
      data-testid="input"
      id="input-22"
      type="textd"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <!---->
      <!---->
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput props label ラベルとして表示 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-label"
  >
    <label
      class="text-sm font-bold py-2"
      data-testid="label"
      for="input-11"
    >
      label
    </label>
    <input
      class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
      data-testid="input"
      id="input-11"
      name="label"
      type="textd"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <!---->
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput props label 空の場合はlabel要素自体を描画しない 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-null"
  >
    <!---->
    <input
      class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
      data-testid="input"
      id="input-12"
      type="textd"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <!---->
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput props rules maxがあれば文字数カウンターを表示 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-null"
  >
    <!---->
    <input
      class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
      data-testid="input"
      id="input-20"
      type="textd"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <div
        class="text-gray-500 text-right mt-1"
        data-testid="counter"
      >
        0 / 300
      </div>
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput props type checkbox 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-null"
  >
    <!---->
    <label
      class="cursor-pointer"
      data-testid="input"
    >
      <input
        class="mr-2"
        type="checkbox"
        value=""
      />
      <span>
        
      </span>
    </label>
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <!---->
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput props type checkboxgroup 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-null"
  >
    <!---->
    <div
      class="flex flex-wrap -mx-4 -my-2"
      data-testid="input"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <!---->
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput props type radio 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-null"
  >
    <!---->
    <div
      class="flex flex-wrap -mx-4 -my-2"
      data-testid="input"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <!---->
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput props type textarea 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-null"
  >
    <!---->
    <textarea
      class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 p-2"
      data-testid="input"
      id="input-10"
      style="height: auto;"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <!---->
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput props value get typeがdatetime-localだった場合は日付文字列にフォーマット 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-null"
  >
    <!---->
    <input
      class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
      data-testid="input"
      id="input-2"
      type="datetime-local"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <!---->
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput props value get v-modelとしてセット 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-null"
  >
    <!---->
    <input
      class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
      data-testid="input"
      id="input-1"
      type="textd"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <!---->
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput renders correctly 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-null"
  >
    <!---->
    <input
      class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
      data-testid="input"
      id="input-0"
      type="textd"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <!---->
    </div>
  </div>
</div>
`;

exports[`components/molecules/AppInput slots label ラベルとして表示 1`] = `
<div>
  <div
    class="flex flex-col untouched pristine"
    data-testid="input-null"
  >
    <label
      class="text-sm font-bold py-2"
      data-testid="label"
      for="input-23"
    >
      label
    </label>
    <input
      class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
      data-testid="input"
      id="input-23"
      type="textd"
    />
    <div
      class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
    >
      <div
        data-testid="errors"
      />
      <!---->
    </div>
  </div>
</div>
`;
